src/orion/algo/asha.py

```diff
@@ -0,0 +1,297 @@
+        """Build rungs based on min_t, max_t, reduction_factor and s.
+
+        :param asha: `ASHA` algorithm
+        :param min_t: Minimum resources (grace_period)
+        :param max_t: Maximum resources
+        :param reduction_factor: Factor of reduction from `min_t` to `max_t`
+        :param s: Minimal early stopping factor (used when there is many brackets)
+        """
+        if min_t <= 0:
+            raise AttributeError("Minimum resources must be a positive number.")
+        elif min_t > max_t:
+            raise AttributeError("Minimum resources must be smaller than maximum resources.")
+
+        self.asha = asha
+        self.reduction_factor = reduction_factor
+        max_rungs = int(numpy.ceil(numpy.log(max_t / min_t) / numpy.log(reduction_factor) - s + 1))
+        self.rungs = [(min(min_t * reduction_factor**(k + s), max_t), dict())
+                      for k in range(max_rungs)]
+
+        logger.debug('Bracket budgets: %s', str([rung[0] for rung in self.rungs]))
+
+    def register(self, point, objective):
+        """Register a point in the corresponding rung"""
+        fidelity = point[self.asha.fidelity_index]
+        rungs = [rung for budget, rung in self.rungs if budget == fidelity]
+        if not rungs:
+            budgets = [budget for budget, rung in self.rungs]
+            raise IndexError(REGISTRATION_ERROR.format(fidelity=fidelity, budgets=budgets,
+                                                       params=point))
+
+        rungs[0][self.asha.get_id(point)] = (objective, point)
+
+    def get_candidate(self, rung_id):
+        """Get a candidate for promotion"""
+        _, rung = self.rungs[rung_id]
+        next_rung = self.rungs[rung_id + 1][1]
+
+        rung = list(sorted((objective, point) for objective, point in rung.values()
+                           if objective is not None))
+        k = len(rung) // self.reduction_factor
+        k = min(k, len(rung))
+
+        for i in range(k):
+            point = rung[i][1]
+            _id = self.asha.get_id(point)
+            if _id not in next_rung:
+                return point
+
+        return None
+
+    @property
+    def is_done(self):
+        """Return True, if reached the bracket reached its maximum resources."""
+        return len(self.rungs[-1][1])
+
+    def update_rungs(self):
+        """Promote the first candidate that is found and return it
+
+        The rungs are iterated over is reversed order, so that high rungs
+        are prioritised for promotions. When a candidate is promoted, the loop is broken and
+        the method returns the promoted point.
+
+        .. note ::
+
+            All trials are part of the rungs, for any state. Only completed trials
+            are eligible for promotion, i.e., only completed trials can be part of top-k.
+            Lookup for promotion in rung l + 1 contains trials of any status.
+
+        """
+        # NOTE: There should be base + 1 rungs
```

Comment:
Is this note still valid after the `reverse` change?

Reply:
Yes but there is no more `base` variable anymore so I'll remove this note indeed.